utils/zerovec/src/map/serde.rs

```diff
@@ -0,0 +1,170 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use super::{ZeroMap, ZeroMapKV, ZeroVecLike};
+use serde::de::{self, MapAccess, Visitor};
+use serde::ser::SerializeMap;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::cmp::Ordering;
+use std::fmt;
+use std::marker::PhantomData;
+
+impl<'a, K, V> Serialize for ZeroMap<'a, K, V>
+where
+    K: ZeroMapKV<'a>,
+    V: ZeroMapKV<'a>,
+    K::Container: Serialize,
+    V::Container: Serialize,
+    K::SerializeType: Serialize,
+    V::SerializeType: Serialize,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        if serializer.is_human_readable() {
+            let mut map = serializer.serialize_map(Some(self.len()))?;
+            for (k, v) in self.iter() {
+                K::get_as_ser(k, |k| map.serialize_key(k))?;
+                V::get_as_ser(v, |v| map.serialize_value(v))?;
+            }
+            map.end()
+        } else {
+            (&self.keys, &self.values).serialize(serializer)
+        }
+    }
+}
+
+/// Modified example from https://serde.rs/deserialize-map.html
+struct ZeroMapMapVisitor<K, V> {
+    marker: PhantomData<fn() -> (K, V)>,
+}
+
+impl<K, V> ZeroMapMapVisitor<K, V> {
+    fn new() -> Self {
+        ZeroMapMapVisitor {
+            marker: PhantomData,
+        }
+    }
+}
+
+impl<'de, K, V> Visitor<'de> for ZeroMapMapVisitor<K, V>
+where
+    K: Deserialize<'de> + Ord,
+    V: Deserialize<'de>,
+    K: ZeroMapKV<'de>,
+    V: ZeroMapKV<'de>,
+{
+    type Value = ZeroMap<'de, K, V>;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("a map produced by LiteMap")
+    }
+
+    fn visit_map<M>(self, mut access: M) -> Result<Self::Value, M::Error>
+    where
+        M: MapAccess<'de>,
+    {
+        let mut map = ZeroMap::with_capacity(access.size_hint().unwrap_or(0));
+
+        // While there are entries remaining in the input, add them
+        // into our map.
+        while let Some((key, value)) = access.next_entry()? {
+            // Try to append it at the end, hoping for a sorted map.
+            // If not sorted, insert as usual.
+            // This allows for arbitrary maps (e.g. from user JSON)
+            // to be deserialized into LiteMap
+            // without impacting performance in the case of deserializing
+            // a serialized map that came from another LiteMap
+            if let Some((key, value)) = map.try_append(key, value) {
+                // Note: this effectively selection sorts the map,
+                // which isn't efficient for large maps
+                map.insert(key, value);
+            }
```

Comment:
For LiteMap, this made sense, but I don't think we should do it for ZeroMap.  We should assume that the ZeroMap is sorted, and eliminate the extra code path.  Our goal isn't to make ZeroMap able to deserialize someone else's map, like it was with LiteMap.

Reply:
Are you sure? If the JSON file is meant to be human readable I assume it's human-editable too, and "keep these keys sorted" seems like a bad idea when it comes to editable  JSON

Reply:
Yes, I think we should be strict with ZeroMap and not assume it can read something which is not sorted.  The burden should be on the person editing JSON to make sure it is sorted.

We shouldn't crash; we should just return a Serde error during deserialization.

Reply:
Okay, will do!